std-bits/table/src/main/java/org/enso/table/data/column/storage/numeric/AbstractLongStorage.java

```diff
@@ -166,13 +207,94 @@ public AbstractLongStorage fillMissingFromPrevious(BoolStorage missingIndicator)
 
   @Override
-  public long get(long index) throws ValueIsNothingException {
-    if (isNothing(index)) {
-      throw new ValueIsNothingException(index);
+  public Storage<Long> applyFilter(BitSet filterMask, int newLength) {
+    var builder =
+        Builder.getForLong(getType(), newLength, BlackholeProblemAggregator.INSTANCE);
+    Context context = Context.getCurrent();
+    for (int i = 0; i < getSize(); i++) {
+      if (filterMask.get(i)) {
+        if (isNothing(i)) {
+          builder.appendNulls(1);
+        } else {
+          builder.appendLong(getItemAsLong(i));
+        }
+      }
+
+      context.safepoint();
+    }
+    return builder.seal();
+  }
+
+  @Override
+  public Storage<Long> applyMask(OrderMask mask) {
+    var builder =
+        Builder.getForLong(getType(), mask.length(), BlackholeProblemAggregator.INSTANCE);
+    Context context = Context.getCurrent();
+    for (int i = 0; i < mask.length(); i++) {
+      int position = mask.get(i);
+      if (position == OrderMask.NOT_FOUND_INDEX || isNothing(position)) {
+        builder.appendNulls(1);
+      } else {
+        builder.appendLong(getItemAsLong(position));
+      }
+
+      context.safepoint();
     }
-    return getItem((int) index);
+    return builder.seal();
   }
 
   @Override
-  public abstract BitSet getIsNothingMap();
+  public Storage<Long> slice(int offset, int limit) {
+    int size = (int) getSize();
+    int newSize = Math.min(size - offset, limit);
+    var builder =
+        Builder.getForLong(getType(), newSize, BlackholeProblemAggregator.INSTANCE);
+    Context context = Context.getCurrent();
+    for (int i = 0; i < newSize; i++) {
+      if (isNothing(offset + i)) {
+        builder.appendNulls(1);
+      } else {
+        builder.appendLong(getItemAsLong(offset + i));
+      }
+      context.safepoint();
+    }
+    return builder.seal();
+  }
+
+  @Override
+  public Storage<Long> slice(List<SliceRange> ranges) {
+    int newSize = SliceRange.totalLength(ranges);
+    var builder =
+        Builder.getForLong(getType(), newSize, BlackholeProblemAggregator.INSTANCE);
+    Context context = Context.getCurrent();
+    for (SliceRange range : ranges) {
+      int rangeStart = range.start();
+      int length = range.end() - rangeStart;
+      for (int i = 0; i < length; i++) {
+        if (isNothing(rangeStart + i)) {
+          builder.appendNulls(1);
+        } else {
+          builder.appendLong(getItemAsLong(rangeStart + i));
+        }
+        context.safepoint();
+      }
+    }
+    return builder.seal();
+  }
+
+  @Override
+  public Storage<Long> appendNulls(int count) {
+    final AbstractLongStorage parent = this;
+    int size = (int) parent.getSize();
+    return new ComputedNullableLongStorage(size + count) {
+      @Override
+      protected Long computeItem(int idx) {
+        if (idx < size) {
+          return parent.getItemBoxed(idx);
+        } else {
+          return null;
+        }
+      }
+    };
+  }
```

Comment:
Not a problem with this PR as this flaw has existed before (so we can create a separate ticket if you prefer), but you made me realize it.

Every call to `appendNulls` creates a new `ComputedNullableLongStorage` layer. In our builders we are very often calling `appendNulls(1)` multiple times. This will create a lot of layers to actually get to the item.

I think we need to keep an overload inside of `ComputedNullableLongStorage` that will ensure such `appendNulls` will be kept flattened - instead of creating new 'layer', it should create a new `ComputedNullableLongStorage` that points to the original parent, thus keeping the 'layering depth' bounded at at most 1.

Essentially same logic as we plan with masks in the future.